snuba/query/joins/semi_joins.py

```diff
@@ -0,0 +1,91 @@
+from typing import Set
+
+from snuba.clickhouse.processors import CompositeQueryProcessor
+from snuba.query import ProcessableQuery
+from snuba.query.composite import CompositeQuery
+from snuba.query.data_source.join import (
+    IndividualNode,
+    JoinClause,
+    JoinModifier,
+    JoinNode,
+    JoinVisitor,
+)
+from snuba.query.data_source.simple import Table
+from snuba.query.expressions import Column
+from snuba.request.request_settings import RequestSettings
+
+
+class SemiJoinOptimizer(CompositeQueryProcessor):
+    """
+    Processes a composite query to find join clause that can be
+    executed as SEMI joins, instead of building the entire join
+    hashtable.
+
+    A semi join is a join where the query only references columns
+    from the left table (it can be used as an INNER join to filter
+    rows of the left table without referencing anything from the
+    right table.
+
+    For such queries the DB does not need to fetch all the rows
+    of the right table that correspond to each row of the left
+    table, thus reducing drastically the memory footprint of the
+    query.
+
+    So this processor navigates the leaves of the join tree from
+    left to right and turn all the join clauses that fit in the
+    conditions above into SEMI joins.
+    As soon as a join clause cannot be transformed we stop there
+
+    """
+
+    def process_query(
+        self, query: CompositeQuery[Table], request_settings: RequestSettings
+    ) -> None:
+        from_clause = query.get_from_clause()
+        if isinstance(from_clause, CompositeQuery):
+            self.process_query(from_clause, request_settings)
+            return
+        elif isinstance(from_clause, ProcessableQuery):
+            return
+
+        # Now this has to be a join, so we can work with it.
+        query.set_from_clause(
+            SemiJoinGenerator(query.get_all_ast_referenced_columns()).visit_join_clause(
+                from_clause
+            )
+        )
+
+
+class SemiJoinGenerator(JoinVisitor[JoinNode[Table], Table]):
+    def __init__(self, referenced_columns: Set[Column]) -> None:
+        self.__referenced_columns = referenced_columns
+
+    def visit_individual_node(
+        self, node: IndividualNode[Table]
+    ) -> IndividualNode[Table]:
+        return node
+
+    def visit_join_clause(self, node: JoinClause[Table]) -> JoinClause[Table]:
+        # TODO: Skip this optimization if the left side of the query
+        # runs aggregations that imply the cardinality of the right
+        # side is important.
+
+        join_cols = set()
+
+        for condition in node.keys:
+            if condition.left.table_alias == node.right_node.alias:
```

Comment:
I'm a little confused by why you check the left table alias. This is here because the left/right of the nodes is not necessarily preserved in the conditions?

Reply:
Exactly. There is nothing in the data structure of the join equality that says that the left table in the condition is actually the left table of the join.
`A INNER JOIN B ON B.x = A.y` is a valid join condition.